### PR TITLE
[JENKINS-34138] Fix maven installs from stepping on each other

### PR DESCRIPTION
**Pull request by t-hall**
When multiple maven projects are launched on a new node the installs can step on each other.  There is locking inside InstallerTranslator but because maven does not have equals/hashcode method and the maven installs are different instances they can run concurrently.  This only works for the JDK because the same instance is used (populated from a global location on the abstract project).  

See [JENKINS-34138](https://issues.jenkins-ci.org/browse/JENKINS-34138).

### Proposed changelog entries

* Entry 1: Fixes maven installs on remote nodes from stepping on each other

### Desired reviewers

@mention
@jenkinsci/code-reviewers

### CODE REVIEW

File: test/src/test/java/hudson/tasks/MavenTest.java [hunk 1]
```diff
@@ -356,4 +354,18 @@ public void parametersReferencedFromPropertiesShouldRetainBackslashes() throws E
         assertTrue("Properties should always be injected, even when build variables injection is enabled",
                 log.contains("-DTEST_PROP1=VAL1") && log.contains("-DTEST_PROP2=VAL2"));
     }
+
+    @Test public void checkMavenInstallationEquals() throws Exception {
```
Comment (oleg-nenashev) [inline test/src/test/java/hudson/tasks/MavenTest.java]: Add `@Issue("JENKINS-34138")`

File: test/src/test/java/hudson/tasks/MavenTest.java [hunk 2]
```diff
@@ -356,4 +354,18 @@ public void parametersReferencedFromPropertiesShouldRetainBackslashes() throws E
...
+        MavenInstallation maven = ToolInstallations.configureMaven3();
+        MavenInstallation maven2 = ToolInstallations.configureMaven3();
+        assertEquals(maven.hashCode(), maven2.hashCode());
+        assertTrue(maven.equals(maven2));
+    }
+
+    @Test public void checkMavenInstallationNotEquals() throws Exception {
```
Comment (oleg-nenashev) [inline test/src/test/java/hudson/tasks/MavenTest.java]: same

File: core/src/main/java/hudson/tasks/Maven.java [hunk 3]
```diff
@@ -709,6 +709,27 @@ public void setInstallations(MavenInstallation... installations) {
...
+            if (getName() != null ? !getName().equals(that.getName()) : that.getName() != null) return false;
+            if (getHome() != null ? !getHome().equals(that.getHome()) : that.getHome() != null) return false;
+            return true;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = getName() != null ? getName().hashCode() : 0;
+            result = 31 * result + (getHome() != null ? getHome().hashCode() : 0);
+            //result = 31 * result + (getProperties() != null ? getProperties().hashCode() : 0);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tasks/Maven.java]: no strong opinion about that. If `getHome()` overlaps, there is likely a problem anyway

**Review by oleg-nenashev — COMMENTED** [review]
The fix LGTM, needs minor patch in the tests

**Review by oleg-nenashev — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers @aheritier WDYT?

**Review by aheritier — COMMENTED** [review]
I'm doubtful @oleg-nenashev 
I agree with you if the Home is the same we are doomed
If the have different names (and potentially different installers/versions we are completely doomed :(
This PR is good (=better than the current state) but we should perhaps start by verifying the home (if set) and then the name for a better behavior. WDYT ?

**Comment by oleg-nenashev** [comment]
> This PR is good (=better than the current state) but we should perhaps start by verifying the home (if set) and then the name for a better behavior. WDYT ?

@t-hall would you be able to update the PR accordingly?

**Comment by t-hall** [comment]
So you are asking to change the order from checking name first to check home first?

**Comment by oleg-nenashev** [comment]
Test failure is unrelated, I'd guess: `testDeleteRecursive_onWindows – hudson.UtilTest`.

**Comment by oleg-nenashev** [comment]
retriggering CI

**Comment by oleg-nenashev** [comment]
@aheritier PTAL

**Comment by oleg-nenashev** [comment]
@aheritier ping. I am going to merge it next week. if there is no negative feedback. Thanks!

**Review by aheritier — APPROVED** [review]
LGTM @oleg-nenashev 
The reorder of imports in MavenTest are probably useless but let's go

**Comment by oleg-nenashev** [comment]
retriggering CI

**Comment by oleg-nenashev** [comment]
Reorder is fine, there is no colliding PRs

**Pull request merged**